### PR TITLE
Other average methods 3d mesh

### PR DESCRIPTION
**Pull request by PeterPetrik**
add all remaining method based on https://github.com/qgis/QGIS-Enhancement-Proposals/issues/158

for methods description see https://fvwiki.tuflow.com/index.php?title=Depth_Averaging_Results

- [X] SingleLevelAverageMethod (top)
- [X] SingleLevelAverageMethod (bottom)
- [X] MultiLevelsFromTopAveragingMethod
- [x] MultiLevelsFromBottomAveragingMethod
- [x] SigmaAveragingMethod
- [x] DepthAveragingMethod
- [x] HeightAveragingMethod
- [x] ElevationAveragingMethod

Screenshot 2019-12-19 at 13 59 47

### CODE REVIEW

File: src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp [hunk 1]
```diff
@@ -49,8 +87,51 @@ std::unique_ptr<QgsMesh3dAveragingMethod> QgsMeshRenderer3dAveragingWidget::aver
...
+      const double startFraction = mSigmaStartFractionSpinBox->value();
+      const double endFraction = mSigmaEndFractionSpinBox->value();
+      averaging.reset( new QgsMeshSigmaAveragingMethod( startFraction, endFraction ) );
+      break;
+    }
+    case QgsMesh3dAveragingMethod::DepthAveragingMethod:
+    {
+      const double startDepth = mDepthStartSpinBox->value();
+      const double endDepth = mDepthEndSpinBox->value();
+      averaging.reset( new QgsMeshMultiLevelsFromBottomAveragingMethod( startDepth, endDepth ) );
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp]: QgsMeshDepthAveragingMethod

File: src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp [hunk 2]
```diff
@@ -49,8 +87,51 @@ std::unique_ptr<QgsMesh3dAveragingMethod> QgsMeshRenderer3dAveragingWidget::aver
...
+      break;
+    }
+    case QgsMesh3dAveragingMethod::HeightAveragingMethod:
+    {
+      const double startHeight = mHeightStartSpinBox->value();
+      const double endHeight = mHeightEndSpinBox->value();
+      averaging.reset( new QgsMeshMultiLevelsFromBottomAveragingMethod( startHeight, endHeight ) );
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp]: QgsMeshHeightAveragingMethod

File: src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp [hunk 3]
```diff
@@ -49,8 +87,51 @@ std::unique_ptr<QgsMesh3dAveragingMethod> QgsMeshRenderer3dAveragingWidget::aver
...
+      break;
+    }
+    case QgsMesh3dAveragingMethod::ElevationAveragingMethod:
+    {
+      const double startVerticalLevel = mElevationStartSpinBox->value();
+      const double endVerticalLevel = mElevationEndSpinBox->value();
+      averaging.reset( new QgsMeshMultiLevelsFromBottomAveragingMethod( startVerticalLevel, endVerticalLevel ) );
```
Comment (wonder-sk) [inline src/app/mesh/qgsmeshrenderer3daveragingwidget.cpp]: QgsMeshElevationAveragingMethod

File: src/core/mesh/qgsmesh3daveraging.h [hunk 4]
```diff
@@ -101,13 +131,26 @@ class CORE_EXPORT QgsMesh3dAveragingMethod
...
+    //! Returns whether the method is correctly initialized
+    virtual bool hasValidInputs() const = 0;
+
+    /**
+     * For one face, calculates absolute vertical levels between which we need to average
+     */
+    virtual void volumeRangeForFace(
+      double &startVerticalLevel,
+      double &endVerticalLevel,
+      const QVector<double> verticalLevels ) const = 0;
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.h]: probably better to use const reference to verticalLevels?

File: src/core/mesh/qgsmesh3daveraging.h [hunk 5]
```diff
@@ -120,17 +163,285 @@ class CORE_EXPORT QgsMeshSingleLevelAveragingMethod: public QgsMesh3dAveragingMe
...
+    double mStartFraction = 0;
+    double mEndFraction = 1;
+};
+
+/**
+ * \ingroup core
+ *
+ * Sigma averages over the values between 0 (bed level) and 1 (surface).
+ *
+ * For example: the average of between a quarter and 3 quarters of the water column - Sigma from 0.25 to 0.75
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.h]: copy-paste error

File: src/core/mesh/qgsmesh3daveraging.h [hunk 6]
```diff
@@ -120,17 +163,285 @@ class CORE_EXPORT QgsMeshSingleLevelAveragingMethod: public QgsMesh3dAveragingMe
...
+    void volumeRangeForFace( double &startVerticalLevel,
+                             double &endVerticalLevel,
+                             const QVector<double> verticalLevels ) const override;
+    double mStartDepth = 0;
+    double mEndDepth = 0;
+};
+
+/**
+ * \ingroup core
+ *
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.h]: missing description

File: src/core/mesh/qgsmesh3daveraging.h [hunk 7]
```diff
@@ -120,17 +163,285 @@ class CORE_EXPORT QgsMeshSingleLevelAveragingMethod: public QgsMesh3dAveragingMe
...
+    void volumeRangeForFace( double &startVerticalLevel,
+                             double &endVerticalLevel,
+                             const QVector<double> verticalLevels ) const override;
+    double mStartHeight = 0;
+    double mEndHeight = 0;
+};
+
+/**
+ * \ingroup core
+ *
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.h]: missing description

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 8]
```diff
@@ -21,6 +21,8 @@
 #include "qgsmeshdataprovider.h"
 #include "qgsmeshrenderersettings.h"
 
+static const double eps = 1e-6;
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: Please explain

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 9]
```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
...
+{
+  return mVerticalLevel >= 1;
+}
+
+void QgsMeshSingleLevelAveragingMethod::volumeRangeForFace(
+  double &startVerticalLevel,
+  double &endVerticalLevel,
+  const QVector<double> verticalLevels ) const
+{
+  if ( ( mVerticalLevel - 1 >= 0 ) && ( mVerticalLevel < verticalLevels.size() ) )
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: ( mVerticalLevel < verticalLevels.size() )  is this correct? shouldn't it be <= if the level is 1..N ?

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 10]
```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
...
+  if ( mStartVerticalLevel > mEndVerticalLevel )
+  {
+    std::swap( mStartVerticalLevel, mEndVerticalLevel );
+  }
+}
+
+QgsMeshMultiLevelsFromTopAveragingMethod::QgsMeshMultiLevelsFromTopAveragingMethod()
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::MultiLevelsFromTopAveragingMethod )
+{
+};
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: extra semicolon

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 11]
```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
...
+}
+
+int QgsMeshMultiLevelsFromTopAveragingMethod::endVerticalLevel() const
+{
+  return mEndVerticalLevel;
+}
+
+bool QgsMeshMultiLevelsFromTopAveragingMethod::hasValidInputs() const
+{
+  return mStartVerticalLevel >= 1;
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: end vertical level probably should be also >= 1 ?

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 12]
```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
...
+  double &endVerticalLevel,
+  const QVector<double> verticalLevels
+) const
+{
+  const double top = verticalLevels[ 0 ];
+  const double bot = verticalLevels[ verticalLevels.size() - 1 ];
+  const double diff = top - bot;
+
+  startVerticalLevel = bot + diff * mStartFraction;
+  endVerticalLevel = bot + diff * mEndFraction;
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: Is this correct? What if vertical levels are not evenly distributed?

File: src/core/mesh/qgsmesh3daveraging.cpp [hunk 13]
```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
...
+}
+
+double QgsMeshElevationAveragingMethod::endElevation() const
+{
+  return mEndElevation;
+}
+
+bool QgsMeshElevationAveragingMethod::hasValidInputs() const
+{
+  return mStartElevation <= 0.0 && mEndElevation <= mStartElevation;
```
Comment (wonder-sk) [inline src/core/mesh/qgsmesh3daveraging.cpp]: I'm not sure if I understand why both of these checks are necessary?

(it is counter-intuitive that end elevation should not be more than start elevation - normally start <= end)

**Review by wonder-sk — COMMENTED** [review]
Generally looking good...

I would suggest to merge 
- MultiLevelsFromTopAveragingMethod with MultiLevelsFromBottomAveragingMethod
- DepthAveragingMethod with HeightAveragingMethod

They are practically the same, there could be just a flag for both indicating whether it is counted from top or bottom. In the user interface they can still appear as separate classes...

File: src/core/mesh/qgsmesh3daveraging.cpp (hunk 9 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmesh3daveraging.cpp]: verticalLevels has size +1 (vertical levels are for all 3d faces so it has size nVolumes + 1 for each face)

File: src/core/mesh/qgsmesh3daveraging.cpp (hunk 11 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmesh3daveraging.cpp]: end is always >= start (see constructor)

File: src/core/mesh/qgsmesh3daveraging.cpp (hunk 12 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmesh3daveraging.cpp]: distribution doesn't matter. startVerticalLevel/endVerticalLevel is in meters

File: src/core/mesh/qgsmesh3daveraging.cpp (hunk 13 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmesh3daveraging.cpp]: we are counting from 0 (surface) down (so the elevation is based on model absolute datum)

**Comment by PeterPetrik** [comment]
@wonder-sk ready for 2nd review

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**